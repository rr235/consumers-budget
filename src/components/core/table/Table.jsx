import React from 'react';
import { arrayOf, shape, string, object, bool, func } from 'prop-types';
import styles from './table.styles.scss';
import { formatCurrencyEUR } from '../../../helper';

const Table = ({ data, keys, onRowClick }) => {
  const getHead = () => keys.map(({ displayName, name }) => <th key={name}>{displayName}</th>);

  const getCol = (row) => {
    return keys.map(({ name, formatCurrencyEuro }) => {
      if (row[name]) {
        let value = row[name];
        if (formatCurrencyEuro) {
          value = formatCurrencyEUR(value);
        }

        return <td key={name}>{value}</td>;
      }
      return null;
    });
  };

  const getRow = () => {
    return data.map((row, index) => (
      // eslint-disable-next-line react/no-array-index-key
      <tr key={index} onClick={() => onRowClick(row)} tabIndex="0">
        {getCol(row)}
      </tr>
    ));
  };

  return (
    <table className={styles.table}>
      <thead>
        <tr>{getHead()}</tr>
      </thead>
      <tbody>{getRow()}</tbody>
    </table>
  );
};

Table.propTypes = {
  keys: arrayOf(
    shape({
      name: string,
      displayName: string,
      formatCurrencyEuro: bool,
    })
  ),
  data: arrayOf(object),
  onRowClick: func,
};

Table.defaultProps = {
  keys: [],
  data: [],
  onRowClick: () => {},
};

export default Table;
