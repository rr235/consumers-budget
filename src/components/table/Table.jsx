import React from 'react';
import { arrayOf, shape, string, object } from 'prop-types';

const Table = ({ data, keys }) => {
  const getHead = () =>
    keys.map(({ displayName, name }) => <th key={name}>{displayName}</th>);

  const getCol = (row) => {
    return keys.map(({ name }) => {
      if (row[name]) {
        return <td key={name}>{row[name]}</td>;
      }
      return null;
    });
  };

  const getRow = () => {
    // eslint-disable-next-line react/no-array-index-key
    return data.map((row, index) => <tr key={index}>{getCol(row)}</tr>);
  };

  return (
    <table>
      <thead>
        <tr>{getHead()}</tr>
      </thead>
      <tbody>{getRow()}</tbody>
    </table>
  );
};

Table.propTypes = {
  keys: arrayOf(
    shape({
      name: string,
      displayName: string,
    })
  ),
  data: arrayOf(object),
};

Table.defaultProps = {
  keys: [],
  data: [],
};

export default Table;
