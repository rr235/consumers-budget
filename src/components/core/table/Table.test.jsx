import React from 'react';
import renderer from 'react-test-renderer';
import Table from './Table';

const data = [
  { col1: 'row1', cole2: 'row1', cole3: 'row1' },
  { col1: 'row2', cole2: 'row2', cole3: 'row2' },
  { col1: 'row3', cole2: 'row3', cole3: 'row3' },
];

const keys = [
  { name: 'col1', displayName: 'Column 1' },
  { name: 'col3', displayName: 'Column 3' },
];

describe('Table', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Table data={data} keys={keys} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
