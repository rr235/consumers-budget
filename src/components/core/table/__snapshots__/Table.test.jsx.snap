// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        Column 1
      </th>
      <th>
        Column 3
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
      tabIndex="0"
    >
      <td>
        row1
      </td>
    </tr>
    <tr
      onClick={[Function]}
      tabIndex="0"
    >
      <td>
        row2
      </td>
    </tr>
    <tr
      onClick={[Function]}
      tabIndex="0"
    >
      <td>
        row3
      </td>
    </tr>
  </tbody>
</table>
`;
